### PR TITLE
[Logging] Added go.uber.org/zap logging service

### PR DESCRIPTION
**Pull request by ale-coinbase**
## Fixes Issue

Fixes: No Production Grade Logging #9
## Changes proposed

1. Added [uber-go/zap](https://github.com/uber-go/zap) logging service to the application

### CODE REVIEW

File: cmd/block_read_poc/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+func NewLogger() (*zap.Logger, error) {
+	var config zap.Config
+	var options []zap.Option
+
+	config = zap.NewDevelopmentConfig()
+	config.EncoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
+	options = append(options, zap.AddStacktrace(zap.FatalLevel))
+
+	return config.Build(options...)
+}
```
Comment (epociask) [inline cmd/block_read_poc/utils/utils.go:18]: Let's move this out of the `cmd` directory and reroute within `/internal`.

File: cmd/block_read_poc/utils/utils.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+// NewLogger ... new logger initialization
+func NewLogger() (*zap.Logger, error) {
+	var config zap.Config
+	var options []zap.Option
+
+	config = zap.NewDevelopmentConfig()
+	config.EncoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
+	options = append(options, zap.AddStacktrace(zap.FatalLevel))
```
Comment (epociask) [inline cmd/block_read_poc/utils/utils.go:15]: Log level might make more sense to have as an environmental variable

File: cmd/block_read_poc/utils/utils.go (hunk 1 above)
Comment (epociask) [reply cmd/block_read_poc/utils/utils.go:18]: Or even a public `/pkg` repo routing that way other developers could import our logger wrapper if needed

Comment (ale-coinbase) [reply cmd/block_read_poc/utils/utils.go:18]: @epociask which would you prefer. I will move to internal for now.

File: internal/config/config.go [hunk 3]
```diff
@@ -25,29 +53,63 @@ type OracleConfig struct {
 }
 
 // NewConfig ... Initializer
-func NewConfig(fileName FilePath) *Config {
+func NewConfig(ctx context.Context, fileName FilePath) *Config {
```
[the remark this answers is not in the document]
Comment (ale-coinbase) [reply internal/config/config.go:56]: nope

File: internal/config/config.go [hunk 4]
```diff
@@ -27,34 +41,75 @@ type OracleConfig struct {
...
 	return envVar
 }
 
-// func convertToInt(str string) int {
-// 	intRep, err := strconv.Atoi(str)
-
-// 	if err != nil {
-// 		panic(err)
-// 	}
+// getEnvBool .. Reads env vars and converts to booleans, panics if incorrect input
```
Comment (epociask) [inline internal/config/config.go:91]:
```suggestion
// getEnvBool ... Reads env vars and converts to booleans, panics if incorrect input
```

File: internal/config/config.go [hunk 5]
```diff
@@ -27,34 +41,75 @@ type OracleConfig struct {
...
+func getEnvBool(key string) bool {
+	if val := getEnvStr(key); val == "1" {
+		return true
+	} else if val == "0" {
+		return false
+	} else {
+		log.Fatalf("env val is not boolean (0 or 1); got: %s=%s", key, val)
```
Comment (epociask) [inline internal/config/config.go:98]: Idk if panicing when a boolean variable isn't 0 or 1 is a correct way to do interpretation here. Typically, you could express `false` by ignoring the env var in your config which this existing logic doesn't currently compensate for. 

Any thoughts? @nadir-akhtar-coinbase

File: config.env.template [hunk 6]
```diff
@@ -1,3 +1,27 @@
...
+ENV=local                               # local,development,production
+
+# Custom Logger Configs 
+LOGGER_USE_CUSTOM=0                     # 0 or 1
+LOGGER_LEVEL=-1                         # -1 (debug), 0 (info), 1 (warn), 2 (error), 3 (dpanic), 4 (panic), 5 (fatal)
+LOGGER_DISABLE_CALLER=0                 # 0 or 1
+LOGGER_DISABLE_STACKTRACE=0             # 0 or 1
+LOGGER_ENCODING=console                 # json,console
+LOGGER_OUTPUT_PATHS=stderr              # comma separated paths
+LOGGER_ERROR_OUTPUT_PATHS=stderr        # comma separated paths
```
Comment (epociask) [inline config.env.template:15]: Awesome job on enumerating all the potential value options

File: internal/config/config.go [hunk 7]
```diff
@@ -27,34 +41,75 @@ type OracleConfig struct {
...
+		L1RpcEndpoint: getEnvStr("L1_RPC_ENDPOINT"),
+		L2RpcEndpoint: getEnvStr("L2_RPC_ENDPOINT"),
+
+		Environment: Env(getEnvStr("ENV")),
+	}
+
+	config.LoggerConfig = &logger.Config{
+		UseCustom:         getEnvBool("LOGGER_USE_CUSTOM"),
+		Level:             getEnvInt("LOGGER_LEVEL"),
+		IsProduction:      config.Environment == Production,
```
[the remark this answers is not in the document]
Comment (ale-coinbase) [reply internal/config/config.go:57]: it starts to get a little funky with import cycles so I have a slightly different approach lmk what you think

File: config.env.template (hunk 6 above)
Comment (ale-coinbase) [reply config.env.template:15]: thx <3

**Review by epociask — APPROVED** [review]
LGTM - Thank you for your contribution sir

**Pull request merged**